### ISSUE TITLE
Issues with URL encoded spaces

### ISSUE THREAD

**Issue by zzxc** [thumbor/thumbor#872]
I read through previous issues with url encoding, but none of the solutions worked in my case.

Here are the test images I'm using:
1. http://belle.zzxc.net/test/nospaces.jpg
2. http://belle.zzxc.net/test/one%20space.jpg

When passing through Thumbor, the percent sign is itself urlencoded, and it stays this way when the image is retrieved from the remote server. This breaks some with some remote servers (eg. nginx, Photobucket), though others (Blogspot) are fine.

It's worth noting curl "http://thumbor-server/unsafe/meta/fit-in/1280x768/http://belle.zzxc.net/test/one space.jpg" works as expected since the space is not urlencoded. This doesn't help in practice, as browsers urlencode the space.

I also tried urlencoding the image path, but that didn't work at all, even on a path without special characters. This would be the ideal solution, as there's no ambiguity in the path.

The problem also affects paths with an encoded plus sign (%2B), parenthesis (%28, %29), and likely other characters. This biggest practical problem is with spaces, as there's no workaround to force a browser to send a literal space in the URL.

### Thumbor request URL
1. http://thumbor-server/unsafe/meta/fit-in/1280x768/http://belle.zzxc.net/test/one%20space.jpg
2. http://thumbor-server/unsafe/meta/fit-in/1280x768/http%3A%2F%2Fbelle.zzxc.net%2Ftest%2Fnospaces.jpg

### Expected behaviour
1. Thumbor retrieves http://belle.zzxc.net/test/one%20space.jpg
2. Thumbor retrieves http://belle.zzxc.net/test/nospaces.jpg

### Actual behaviour
1. Thumbor retrieves http://belle.zzxc.net/test/one%2520space.jpg, resulting in a 404.

When computing the SHA1 hash for security, the URL it tries to use is: http://belle.zzxc.net/test/one%252520space , where the % symbol has been url encoded twice.

Console log:
```

2017-02-27 04:03:13 thumbor:WARNING ERROR retrieving image http://belle.zzxc.net/test/one%2520space: HTTP 404: Not Found
2017-02-27 04:03:13 tornado.access:WARNING 404 GET /unsafe/fit-in/1024x768/http://belle.zzxc.net/test/one%252520space (65.24.37.221) 109.78ms
```

Remote nginx log:
a.b.c.d - - [27/Feb/2017:04:03:13 +0000] "GET /test/one%2520space HTTP/1.1" 404 152 "-" "Thumbor/6.2.1"

2. A 400 error is returned to the browser, and this warning is on the console:

```
2017-02-27 04:02:11 thumbor:WARNING Request denied because the specified path "http%25253A%25252F%25252Fbelle.zzxc.net%25252Ftest%25252Fnospaces.jpg" was not identified by the loader as a valid path
2017-02-27 04:02:11 thumbor:WARNING No original image was specified in the given URL
2017-02-27 04:02:11 tornado.access:WARNING 400 GET /unsafe/fit-in/1024x768/http%25253A%25252F%25252Fbelle.zzxc.net%25252Ftest%25252Fnospaces.jpg (65.24.37.221) 2.13ms
```

### Operating system
Ubuntu 16.04

### Your thumbor.conf
PNG_COMPRESSION_LEVEL = 8
RESULT_STORAGE = 'thumbor.result_storages.file_storage'
SECURITY_KEY = 'secret'
#ALLOW_UNSAFE_URL = False
ALLOW_OLD_URLS = False
STORAGE_EXPIRATION_SECONDS = 3154000000 #100 years
FILE_STORAGE_ROOT_PATH = '/home/thumbor/thumborcache/storage'

**Comment by zzxc** [comment]
This was bothering me, so I briefly looked into the code. Here's what I did to get rid of redundant urlencoding:

`diff --git a/thumbor/handlers/imaging.py b/thumbor/handlers/imaging.py
index 30c3431..de4394c 100644
--- a/thumbor/handlers/imaging.py
+++ b/thumbor/handlers/imaging.py
@@ -34,7 +34,8 @@ class ImagingHandler(ContextHandler):
 
         url = self.request.path
 
-        kw['image'] = quote(kw['image'].encode('utf-8'))
+        #This has already been urlencoded by Tornado; let's not do it twice.
+        #kw['image'] = quote(kw['image'].encode('utf-8'))
         if not self.validate(kw['image']):
             self._error(400, 'No original image was specified in the given URL')
             return
`

To make the string to generate the HMAC hash, I had to double urlencode the image URL. I excluded the characters /, : = and & from urlencoding and (due to separate processing of the query string) only urlencoded the ? character once.

Remote URL: 
Thumbor URL: https://thumbor-server/w1HuDdbEksNLAVtBsGiLndPpKU8=/fit-in/1280x768/http://belle.zzxc.net/test/one%20space.jpg?query=string
String for HMAC:

**Comment by masom** [comment]
@zzxc A big reason this is happening is because the URL contains "invalid" characters. the `http://...` portion of the source URL should be urlencoded before being HMAC'ed.

Your url to thumbor should look like this: `http://thumbor-server/<signature>/fit-in/http%3A%2F%2Fbelle.zzxc.net%2Ftest%2Fone%2520space.jpg%3Fquery%3Dstring`

**Comment by zzxc** [comment]
So, if I'm understanding correctly, the urlencoded example should work with a valid HMAC hash, but not with the "unsafe" token?

I tried an example like yours in my first attempt to work around this issue. The result was as mentioned above as the second URL under Request URL and Actual Behavior. Granted, this was without HMAC.

**Comment by masom** [comment]
With/without HMAC the URL to the source image should always be urlencoded, otherwise browsers/clients might act weirdly. The source URL is part of the URI path.

**Comment by zzxc** [comment]
@masom I tried this URL in Firefox and Chrome:

https://thumbor-server/LHvQCUKb61nX5iFHTMei5k3BMj8=/fit-in/1280x768/http%3A%2F%2Fbelle.zzxc.net%2Ftest%2Fone%2520space.jpg%3Fquery%3Dstring

The browser again gets an http 400, with this on the console:
```

2017-02-27 15:38:07 thumbor:WARNING Request denied because the specified path "http%253A%252F%252Fbelle.zzxc.net%252Ftest%252Fone%252520space.jpg%253Fquery%253Dstring" was not identified by the loader as a valid path
2017-02-27 15:38:07 thumbor:WARNING No original image was specified in the given URL
2017-02-27 15:38:07 tornado.access:WARNING 400 GET /LHvQCUKb61nX5iFHTMei5k3BMj8=/fit-in/1280x768/http%25253A%25252F%25252Fbelle.zzxc.net%25252Ftest%25252Fone%25252520space.jpg%25253Fquery%25253Dstring (65.24.37.221) 1.75ms

```

Is there something I need to be doing so that it parses urlencoded image paths? I am using the most recent git master revision of Thumbor (with and without the change in my first comment), and dependencies were installed with pip.

**Comment by mpilar** [comment]
You are double encoding when you send the encoded space to be encoded. Ie. %20 is "space" encoded. Your URL is either `http://belle.zzxc.net/test/one space.jpg` (which will act weird on some browsers) or `http%3A%2F%2Fbelle.zzxc.net%2Ftest%2Fone%20space.jpg` (which should work correctly)

**Comment by zzxc** [comment]
Thanks for the pointers; everything works as expected when urlencoded exactly once, so this is not an issue with Thumbor.

The issue I was running into with the fully encoded paths was that nginx redundantly escapes and/or unescapes the path when transforming it prior to the proxy_pass directive. I was transforming the path to put Thumbor on the same hostname as other parts of the site, at /thumbor/<thumbor-path>. I did it this way since the site owner hadn't purchased a wildcard SSL certificate.

The long term fix is to set up a separate hostname for Thumbor; this is needed anyway to avoid the problem of cookies hitting an image cache. I fixed it for now with an ugly nginx configuration change. (https://www.ruby-forum.com/topic/4413745)

Supporting nginx-normalized paths in Thumbor would be ideal, but getting this right without breaking hashes would require very specific assumptions about the urlencoding applied.
